### PR TITLE
libvirt_mem: Added support for multiple memory hotplug

### PR DESCRIPTION
**Pull request by nasastry**
Added support for handling multiple memory hotplug requests

Signed-off-by: Nageswara R Sastry rnsastry@linux.vnet.ibm.com

### CODE REVIEW

File: libvirt/tests/src/libvirt_mem.py [hunk 1]
```diff
@@ -99,14 +99,23 @@ def check_guest_meminfo(old_mem):
...
         new_mem = get_vm_memtotal(session)
         session.close()
         logging.debug("Memtotal on guest: %s", new_mem)
-        if new_mem != int(old_mem) + int(tg_size):
-            raise error.TestFail("Total memory on guest couldn't"
-                                 " changed after attach memory "
-                                 "device")
+        no_of_times = 1
+        if at_times:
+            no_of_times = int(at_times)
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_mem.py]: at_times is already int type during assignment, can be used directly in all other occurrences.

Comment (nasastry) [inline libvirt/tests/src/libvirt_mem.py]: When attaching single memory device 'at_times' not used in this situation need to assign 'no_of_times' to '1'.

Comment (balamuruhans) [inline libvirt/tests/src/libvirt_mem.py]: No, type int(at_times) is not required as it is already int type during assignment "at_times = int(params.get("attach_times", 1))"

Comment (will-Do) [inline libvirt/tests/src/libvirt_mem.py]: @nasastry 
I guess @balamuruhans means you don't need to convert ```at_times``` to int as it already done in line 326

Comment (nasastry) [inline libvirt/tests/src/libvirt_mem.py]: @will-Do, @balamuruhans Done the changes and pushed the new modified code.

**Comment by will-Do** [comment]
Just run the test, and one case failed:
```
 (03/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.cold_plug: FAIL (51.92 s)

2016-09-28 11:37:18,705 stacktrace       L0039 ERROR| Reproduced traceback from: /tmp/libvirt-ci-debug/avocado-vt/avocado_vt/test.py:420
2016-09-28 11:37:18,705 stacktrace       L0042 ERROR| Traceback (most recent call last):
2016-09-28 11:37:18,705 stacktrace       L0042 ERROR|   File "/tmp/libvirt-ci-debug/avocado-vt/avocado_vt/test.py", line 206, in runTest
2016-09-28 11:37:18,706 stacktrace       L0042 ERROR|     raise exceptions.TestFail(details)
2016-09-28 11:37:18,706 stacktrace       L0042 ERROR| TestFail: Found unmatched memory setting from domain xml

**Comment by nasastry** [comment]
@will-Do 
I ran many times, but didn't see this error. Any way I am looking in to it. Thanks for the feedback.

**Comment by nasastry** [comment]
@will-Do 
I ran this on ppc64le arch, with a customized (libvirt_mem.cfg).
I guess you are ran on x86_64 arch. Can you please share the test cfg file (not the libvirt_mem.cfg) by that I will try to re-produce the above error.
Thanks in advance.

**Comment by will-Do** [comment]
@nasastry 
Yes, I'm on x86_64, and with all cfg files as default.

File: libvirt/tests/src/libvirt_mem.py [hunk 2]
```diff
@@ -99,14 +99,23 @@ def check_guest_meminfo(old_mem):
...
-        if new_mem != int(old_mem) + int(tg_size):
-            raise error.TestFail("Total memory on guest couldn't"
-                                 " changed after attach memory "
-                                 "device")
+        no_of_times = 1
+        if at_times:
+            no_of_times = at_times
+        if attach_device:
+            if new_mem != int(old_mem) + (int(tg_size) * no_of_times):
+                raise error.TestFail("Total memory on guest couldn't"
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_mem.py]: error.TestFail can be replaced with exceptions.TestFail

File: libvirt/tests/src/libvirt_mem.py [hunk 3]
```diff
@@ -99,14 +99,23 @@ def check_guest_meminfo(old_mem):
...
+                                     " changed after attach memory "
+                                     "device")
+        if detach_device:
+            if new_mem != int(old_mem) - (int(tg_size) * no_of_times):
+                raise error.TestFail("Total memory on guest couldn't"
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_mem.py]: same here

**Comment by balamuruhans** [comment]
@nasastry failure "Found unmatched memory setting from domain xml" might be due to unsupported libvirt version, memory hotplug support QEMU/KVM driver was added in 1.2.14 version of libvirt. 

Test can be skipped with this check.

**Comment by nasastry** [comment]
@will-Do, @balamuruhans 
Done with the suggested changes.
Added libvirt version check and changed error to exceptions.
Thanks!!

**Comment by will-Do** [comment]
@nasastry 
Will check this later, thx.

**Comment by will-Do** [comment]
@nasastry 
Hi, just run there test, and still got 2 failures, here is the result:
```
# rpm -q libvirt
libvirt-2.0.0-10.el7.x86_64

# avocado run --vt-type libvirt  libvirt_mem
JOB ID     : 386f8904e563197ec3c979068ec8ea5c93d05a68
JOB LOG    : /root/avocado/job-results/job-2016-11-21T14.25-386f890/job.log
TESTS      : 22
 (01/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.mem_basic: PASS (54.22 s)
 (02/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.hot_plug: PASS (68.77 s)
 (03/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.cold_plug: FAIL (51.92 s)
 (04/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.no_attach: PASS (53.00 s)
 (05/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.with_source: PASS (51.95 s)
 (06/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.hugepages.no_numatune.hot_plug: PASS (215.97 s)
 (07/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.hugepages.no_numatune.cold_plug: PASS (199.50 s)
 (08/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.hugepages.with_numatune.hot_plug: PASS (216.20 s)
 (09/22) type_specific.io-github-autotest-libvirt.libvirt_mem.positive_test.hugepages.with_numatune.cold_plug: PASS (199.35 s)
 (10/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.start_error.invalid_slot: PASS (3.67 s)
 (11/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.start_error.invalid_address: PASS (3.93 s)
 (12/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.start_error.invalid_vm_node: PASS (3.48 s)
 (13/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.start_error.invalid_host_node: PASS (3.45 s)
 (14/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.start_error.without_numa: PASS (3.42 s)
 (15/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.attach_error.attach_nonexist_node: PASS (21.59 s)
 (16/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.attach_error.attach_invalid_type: PASS (21.82 s)
 (17/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.attach_error.attach_duplicate_node: PASS (21.89 s)
 (18/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.attach_error.attach_many_times: PASS (22.05 s)
 (19/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.attach_error.attach_invalid_size: FAIL (21.86 s)
 (20/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.detach_error: PASS (21.69 s)
 (21/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.setmem_error.without_numa: PASS (2.73 s)
 (22/22) type_specific.io-github-autotest-libvirt.libvirt_mem.negative_test.setmem_error.with_numa: PASS (31.33 s)
RESULTS    : PASS 20 | ERROR 0 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2016-11-21T14.25-386f890/html/results.html
TESTS TIME : 1293.77 s

**Comment by nasastry** [comment]
@will-Do, I will look in to them. Thanks!!

**Comment by waynesun09** [comment]
https://github.com/autotest/tp-libvirt/pull/944 fixed the invalid_size failure

**Comment by nasastry** [comment]
@waynesun09 mentioned invalid_size failures got fixed.
@will-Do can you please pull this request?

**Pull request merged**